drivers/include/mfrc522.h

```diff
@@ -0,0 +1,844 @@
+ */
+
+/**
+ * @defgroup drivers_mfrc522 MFRC522 controller
+ * @ingroup  drivers_actuators
+ * @brief    Device driver for the NXP MFRC522
+ *
+ * Data sheet available here: https://www.nxp.com/docs/en/data-sheet/MFRC522.pdf
+ *
+ * This driver is a port of a driver originally written for Arduino, you can find
+ * the original here: https://github.com/miguelbalboa/rfid (currently based on
+ * commit 0a568b45baf1852883630e90ea125786e88f5322)
+ *
+ * @{
+ *
+ * @author   Hendrik van Essen <hendrik.ve@fu-berlin.de>
+ * @file
+ */
+
+#ifndef MFRC522_H
+#define MFRC522_H
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "stdbool.h"
+
+#include "periph/spi.h"
+
+#include "mfrc522_regs.h"
+
+/**
+ * @brief MFRC522 device initialization parameters
+ */
+typedef struct {
+    spi_t spi_dev;      /**< SPI bus the controller is connected to */
+    spi_clk_t spi_clk;  /**< SPI clock speed */
+    gpio_t sck_pin;     /**< SCK pin */
+    gpio_t miso_pin;    /**< MISO pin */
+    gpio_t mosi_pin;    /**< MOSI pin */
+    gpio_t cs_pin;      /**< Slave select pin */
+    gpio_t rst_pin;     /**< Reset pin */
+} mfrc522_params_t;
+
+/**
+ * @brief MFRC522 device data structure type
+ */
+typedef struct {
+    mfrc522_params_t params;  /**< Device initialization parameters */
+} mfrc522_t;
+
+/**
+ * @brief A struct used for passing the UID of a PICC
+ */
+typedef struct {
+    uint8_t size;          /**< Number of bytes in the UID. 4, 7 or 10 */
+    uint8_t uid_byte[10];  /**< UID */
+    uint8_t sak;           /**< The SAK (Select acknowledge) byte returned from
+                                the PICC after successful selection */
+} mfrc522_uid_t;
+
+/**
+ * @brief MIFARE constants that does not fit anywhere else
+ */
+typedef enum {
+    MFRC522_MF_ACK      = 0xA,  /**< The MIFARE Classic uses a 4 bit ACK/NAK.
+                                     Any other value than 0xA is NAK */
+    MFRC522_MF_KEY_SIZE = 6     /**< A Mifare Crypto1 key is 6 bytes */
+} mfrc522_mifare_misc_t;
+
+/**
+ * @brief A struct used for passing a MIFARE Crypto1 key
+ */
+typedef struct {
+    uint8_t key_byte[MFRC522_MF_KEY_SIZE];  /**< Crypto1 key bytes */
+} mfrc522_mifare_key_t;
+
+/**
+ * @brief MFRC522 status codes
+ */
+typedef enum {
+    MFRC522_STATUS_OK,              /**< Success */
+    MFRC522_STATUS_ERROR,           /**< Error in communication */
+    MFRC522_STATUS_COLLISION,       /**< Collision detected */
+    MFRC522_STATUS_TIMEOUT,         /**< Timeout in communication */
+    MFRC522_STATUS_NO_ROOM,         /**< A buffer is not big enough */
+    MFRC522_STATUS_INTERNAL_ERROR,  /**< Internal error in the code */
+    MFRC522_STATUS_INVALID,         /**< Invalid argument */
+    MFRC522_STATUS_CRC_WRONG,       /**< The CRC_A does not match */
+    MFRC522_STATUS_MIFARE_NACK,     /**< A MIFARE PICC responded with NAK */
+    MFRC522_STATUS_UNKNOWN          /**< Status used for mfrc522_status_code_names */
+} mfrc522_status_code_t;
```

Comment:
Please drop this type and use errno codes for error reporting. Refer to the driver programming guide on why standard error codes are preferred over custom codes for every driver.